### ISSUE TITLE
pytorch frozen parameter training failed

### ISSUE THREAD

**Issue by AndyYuan96** [horovod/horovod#1618]
**Environment:**
1. Framework: PyTorch
Framework version: 1.1.0
Horovod version: 0.18.2
MPI version: openmpi-4.0.2
CUDA version: 10.0
NCCL version: 2.5.6
Python version: 3.6
OS and version: Ubuntu
GCC version: 5.4.0

**Checklist:**
1. Did you search issues to find if somebody asked this question before? nobody.
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Please ask your question here.
Hi, I want to training a network, and frozen some parameter, only training the parameter whose requires_grad = True.

So I first set the frozen parameter's requires_grad = False, and other's to True.
After finish model parameter init:

1. broadcast init parameters to all gpu, including frozened parameters and training parameters.
hvd.broadcast_parameters(self.model.state_dict(), root_rank=0)

2. init pytorch optimizer: I only use the requires_grad=True 's parameter.
params_dict = [ {'params': filter(lambda p: p.requires_grad, self.model.parameters())}]
self.optimizer = torch.optim.Adam(
            params=params_dict,
            lr=self.settings.lr,
            weight_decay=self.settings.weight_decay,
        )

3. wrap pytorch optimizer:  I only use the requires_grad=True 's parameter
self.optimizer = hvd.DistributedOptimizer(
            self.optimizer,
            named_parameters=chain(filter(lambda p: p[1].requires_grad, model.named_parameters())))

4. hvd.broadcast_optimizer_state(self.optimizer, root_rank=0)

5. 
self.optimizer.zero_grad()
loss.backward()
self.optimizer.step()

when training with multi gpu, the error is below:

[2019-12-27 08:09:44.346946: W horovod/common/operations.cc:779] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, whi
ch will cause deadlock. 
Stalled ranks:
0: [allreduce.bev_conv.0.bias, allreduce.bev_conv.0.weight, allreduce.bev_conv.2.bias, allreduce.bev_conv.2.weight, allreduce.bev_conv.4.bias, allreduce.bev_conv.4.weight ...]
[2019-12-27 08:10:44.349347: W horovod/common/operations.cc:779] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, whi
ch will cause deadlock. 
Stalled ranks:
0: [allreduce.bev_conv.0.bias, allreduce.bev_conv.0.weight, allreduce.bev_conv.2.bias, allreduce.bev_conv.2.weight, allreduce.bev_conv.4.bias, allreduce.bev_conv.4.weight ...]
1: [allreduce.block1.0.bias, allreduce.block1.0.weight, allreduce.block1.2.bias, allreduce.block1.2.weight, allreduce.block1.4.bias, allreduce.block1.4.weight ...]
[2019-12-27 08:11:44.353990: W horovod/common/operations.cc:779] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, whi
ch will cause deadlock. 

Explain:
allreduce.bev_conv.0.bias, allreduce.bev_conv.0.weight, allreduce.bev_conv.2.bias, allreduce.bev_conv.2.weight, allreduce.bev_conv.4.bias, allreduce.bev_conv.4.weight  are my frozened parameter's name, it's really sucks, I don't why that I didn't  pass those tensor to pytorch optimizer and DistributedOptimizer, but horvod still reduce them?

allreduce.block1.0.bias, allreduce.block1.0.weight, allreduce.block1.2.bias, allreduce.block1.2.weight, allreduce.block1.4.bias, allreduce.block1.4.weight  are my training parameter's name.

Does horovod only support training the whole network's parameter with multi gpu? when using one gpu, no error.

**Comment by AndyYuan96** [comment]
I find the reason , but don't know how to solve.
I use two gpus,
when init pytorch optimizer:
params_dict = [ {'params': filter(lambda p: p.requires_grad, self.model.parameters())}]
self.optimizer = torch.optim.Adam(
params=params_dict,
lr=self.settings.lr,
weight_decay=self.settings.weight_decay,
)
before I init pytorch optimizer, I print the parameter that has grad:
print("current rank is : ", hvd.rank())
for p in filter(lambda p: p.requires_grad, self.model.parameters()):
        if p.requires_grad:
            print(p.shape)
the result show that rank 0 is parameter is right, but rank 1 print the whole network's parameters.
It's so wired,  can you tell me what I can do?

**Comment by tgaddair** [comment]
Hey @AndyYuan96, if rank 1 is printing the whole network's parameters before you construct the optimizer, I'm wondering if it also does so before you call `hvd.broadcast_parameters(...)`.  It may be that `broadcast_parameters` is not working correctly with your model for some reason.

I tried reproducing this, but was not able to get this behavior.  Do you have a small script you can share that will demonstrate this problem?

**Comment by AndyYuan96** [comment]
@tgaddair , fixed, my mistake, I only set parameter's requries_grad in rank0.

**Comment by AndyYuan96** [comment]
@tgaddair , fixed, my mistake, I only set parameter's requries_grad in rank0.
